process/block/metablock.go

```diff
@@ -1888,8 +1946,14 @@ func (mp *metaProcessor) checkShardHeadersFinality(
 		// verify if there are "K" block after current to make this one final
 		nextBlocksVerified := uint32(0)
 		for _, shardHdr := range finalityAttestingShardHdrs[shardId] {
-			err := mp.checkShardHeaderFinalityBasedOnProofs(shardHdr, shardId)
-			if err != nil {
+			isNotarizedBasedOnProofs, errCheckProof := mp.checkShardHeaderFinalityBasedOnProofs(shardHdr, shardId)
+			// if the header is notarized based on proofs and there is no error, break the loop
+			// if there is any error, forward it, header is not final
+			if isNotarizedBasedOnProofs {
+				if errCheckProof != nil {
+					return err
+				}
+
```

Comment:
the breaking from the loop at line 1948 should also invalidate the condition at 1978, and not request any blocks as the block does not need any confirmation in this case.